include/verilated_vpi.cpp

```diff
@@ -326,21 +355,52 @@ class VerilatedVpioVar VL_NOT_FINAL : public VerilatedVpioVarBase {
     static VerilatedVpioVar* castp(vpiHandle h) {
         return dynamic_cast<VerilatedVpioVar*>(reinterpret_cast<VerilatedVpio*>(h));
     }
+    uint32_t bitOffset() const override { return m_bitOffset; }
     uint32_t mask() const { return m_mask.u32; }
     uint8_t mask_byte(int idx) const { return m_mask.u8[idx & 3]; }
     uint32_t entSize() const { return m_entSize; }
-    uint32_t index() const { return m_index; }
+    const std::vector<int32_t>& index() const { return m_index; }
+    VerilatedVpioVar* withIndex(int32_t index) const {
+        if (VL_UNLIKELY(indexedDim() + 1 >= varp()->dims()))
+            return nullptr;
+
+        auto ret = new VerilatedVpioVar{this};
+        ret->m_index.push_back(index);
+        ret->m_indexedDim++;
+
+        int chunkSize = 1;
+        for (int dim = maxDim(isIndexedDimUnpacked()); dim > indexedDim() + 1; dim--)
+            chunkSize *= varp()->range(dim)->elements();
+
+        if (isIndexedDimUnpacked())
+            ret->m_varDatap = (static_cast<uint8_t*>(ret->m_varDatap)) + entSize() * chunkSize * (index - get_range()->low());
+        else
+            ret->m_bitOffset += chunkSize * (index - get_range()->low());
+
+        return ret;
+    }
     uint32_t type() const override {
         uint32_t type = vpiReg;
         switch (varp()->vltype()) {
         case VLVT_REAL: type = vpiRealVar; break;
         case VLVT_STRING: type = vpiStringVar; break;
         default: break;
         }
-        return (varp()->dims() > 1) ? vpiMemory : type;  // but might be wire, logic
+        if (isIndexedDimUnpacked())
+            return vpiRegArray;
+        else
+            return type;
+    }
+    const char* fullname() const override {
```

Comment:
I got undefined behavior in my code because all other `fullname` functions return pointer to a member, while this function points to a static. It's quite confusing.
Can we refactor the code to generate full name in the constructor?

Reply:
This pull is closed, if you get warnings on this please open a new issue.   Note the fullname is already in the constructor, and we return strings similar elsewhere, so a different fix will be needed.

Reply:
Note IEEE 38.11 "The string shall be placed in a temporary buffer that shall be used by every call to this routine. If the string is to be used after a subsequent call, the string should be copied to another location.

Reply:
Ah, I see. So technically current behavior is up to the standard :thinking: 
So do you think that other functions need to be adjusted to use an intermediate buffer instead?

Reply:
If writing VPI code, I'd probably be safe and assume everything needs to be copied.

Reply:
In my case I use VPI just to interact with Verilator, I don't plan it to be portable...
I could save strings, but since Verilator already stores them (for everything except array pins), I tried to reuse the memory. It's especially handy for big designs where simulation already takes a lot of RAM.
If we adjust the logic for array pins to store the name inside as we do for cells and modules, will it violate the standard? Technically we just making the pointer lifetime longer, so I think it's backwards compatible?

Reply:
We'd need RAM for every index, I don't see that as reasonable, so no change please.

Reply:
Makes sense. Thanks for the quick response!